ENIGMAsystem/SHELL/Universal_System/Extensions/ProcInfo/Win32/procinfo.cpp

```diff
@@ -0,0 +1,451 @@
+  tfn_qip pfn_qip = tfn_qip(GetProcAddress(p_ntdll, "NtQueryInformationProcess"));
+  PROCESS_BASIC_INFORMATION pbi;
+  ULONG res_len = 0;
+  NTSTATUS status = pfn_qip(hproc, ProcessBasicInformation, &pbi, sizeof(pbi), &res_len);
+  size_t ppeb = size_t(pbi.PebBaseAddress);
+  char peb[sizeof(PEB)];
+  SIZE_T read;
+  ReadProcessMemory(hproc, pbi.PebBaseAddress, peb, sizeof(peb), &read); 
+  enum { OFFSET_PEB = 0x10, OFFSET_X = 0x48 };
+  void *ptr = (void *) *(INT_PTR *)(peb + OFFSET_PEB);
+  char buffer[OFFSET_X + sizeof(void *)];
+  ReadProcessMemory(hproc, ptr, buffer, sizeof(buffer), &read); 
+  void *penv = (void *) *(INT_PTR *)(buffer + OFFSET_X);
+  SIZE_T nSiZe;
+  if (!HasReadAccess(hproc, penv, nSiZe)) { return (wchar_t *)""; }
+  wchar_t *env = new wchar_t[nSiZe];
+  ReadProcessMemory(hproc, penv, env, nSiZe, &read); 
+  return env;
+}
+
+static string proc_wids;
+static process_t proc_pid;
+static inline BOOL CALLBACK EnumWindowsProc(HWND hWnd, LPARAM lparam) {
+  DWORD dw_pid;
+  GetWindowThreadProcessId(hWnd, &dw_pid);
+  if (dw_pid == proc_pid) {
+    void *voidp_wid = reinterpret_cast<void *>(hWnd);
+    proc_wids += to_string(reinterpret_cast<unsigned long long>(voidp_wid));
+    proc_wids += "|";
+  }
+  return TRUE;
+}
+
+namespace procinfo {
+
+static process_t prevpid;
+static string prevout;
+
+process_t process_execute(string command) {
+  process_t pid = 0;
+  string output;
+  wstring wstr_command = widen(command);
+  wchar_t cwstr_command[32768];
+  wcsncpy_s(cwstr_command, 32768, wstr_command.c_str(), 32768);
+  BOOL proceed = TRUE;
+  HANDLE hStdInPipeRead = NULL;
+  HANDLE hStdInPipeWrite = NULL;
+  HANDLE hStdOutPipeRead = NULL;
+  HANDLE hStdOutPipeWrite = NULL;
+  SECURITY_ATTRIBUTES sa = { sizeof(SECURITY_ATTRIBUTES), NULL, TRUE };
+  proceed = CreatePipe(&hStdInPipeRead, &hStdInPipeWrite, &sa, 0);
+  if (proceed == FALSE) return pid;
+  proceed = CreatePipe(&hStdOutPipeRead, &hStdOutPipeWrite, &sa, 0);
+  if (proceed == FALSE) return pid;
+  STARTUPINFOW si = { 0 };
+  si.cb = sizeof(STARTUPINFOW);
+  si.dwFlags = STARTF_USESTDHANDLES;
+  si.hStdError = hStdOutPipeWrite;
+  si.hStdOutput = hStdOutPipeWrite;
+  si.hStdInput = hStdInPipeRead;
+  PROCESS_INFORMATION pi = { 0 };
+  DWORD dwRead = 0;
+  DWORD dwAvail = 0;
+  if (CreateProcessW(NULL, cwstr_command, NULL, NULL, TRUE, CREATE_NO_WINDOW, NULL, NULL, &si, &pi)) {
+    while (WaitForSingleObject(pi.hProcess, 5) == WAIT_TIMEOUT) {
+      PeekNamedPipe(hStdOutPipeRead, NULL, 0, NULL, &dwAvail, NULL);
+      if (dwAvail) {
+        vector<char> buffer(dwAvail);
+        BOOL success = ReadFile(hStdOutPipeRead, &buffer[0], dwAvail, &dwRead, NULL);
+        if (success || dwRead) {
+          buffer[dwAvail] = 0;
+          output.append(buffer.data(), dwAvail);
+        }
+      }
+      MSG msg;
+      if (PeekMessage(&msg, NULL, 0, 0, PM_REMOVE)) {
+        TranslateMessage(&msg);
+        DispatchMessage(&msg);
+      }

```

Comment:
`enigma::handleEvents()`

Reply:
Again, this extension I need to be able to use the source code without being dependent on the rest of enigma; This pull request I am not expecting to actually have merged into master, I just made it for code review purposes and for those who care enough to use it in enigma can download it separately because Josh doesn't want this in enigma by default. Sorry for the confusion, I am also building from this exact code as a GM extension so it needs to not have dependency on enigma's main source.

Reply:
But i can copy/past the code from handleEvents if i were allowed to relicense it under MIT which i've done with a lot of enigma's code I've written and i don't want to make the GPL enigma has go to shit.

Reply:
Helper

Reply:
Oh, I see, you just want an honest code review. Alright, with everything I learned about processes from adding the stop button, I will try to catch any bugs in your code.

Reply:
Thank you so much Robert! You, Josh, fundies, and the programming community is generally just very helpful without necessarily expecting any personal gain, and it's a joy to work with.

Reply:
For this one, it actually doesn't matter, you should still make a helper. It should be using a helper in your own version in your own repo. If you did that you would only need to delete your helper when bringing it into. And to make that even easier, you could put all of the helpers you create, which you know ENIGMA will, into a single file, so when you update ENIGMA, you'll only have to delete a single file.

Reply:
That's actually a really good idea. I could define a macro in my dll build scripts, since those are not downloaded with the rest of the repository, and when do an #ifdef for the macro for including the helper header, that way the helpers will only be built into the GM dll's and not into the enigma game exe's.